### PR TITLE
Create production container

### LINKED ISSUE #45: Create a Production Docker container
- Use Datagateway as an example (https://github.com/ral-facilities/datagateway/pull/1535)

### PR DESCRIPTION
**Pull request by VKTB**
## Description
This PR adds a `Dockerfile.prod` which can be used to build a Docker image and create a container using it. It is configured to create a production image and run an Apache httpd server on port `80` when a container is started.

To add an extra layer of security, the image has been set up so that containers run as non-root.

Environment variables have also been defined in the Dockerfile to allow for values to be passed at runtime to future running containers. These values are then used by the `docker-entrypoint.sh` script, which acts as an entry point script that is executed when the container starts, to update the config values in the `nventory-management-system-settings.example.json1 file. The environment variables are:
- `API_URL` (Default value: `/inventory-management-system-api`)

Finally, this PR adds a `docker` job to the `ci-build` GitHub Actions workflow to build a Docker image using the Dockerfile introduced in this PR and push the same to Harbor.

To build an image, run the command below in the `inventory-management-system` repo directory
```
docker build -t inventory_management_system_image -f Dockerfile.prod .
```
To start a container on port `80` from the image that you just built, run:
```
docker run -p 80:80 --name inventory_management_system_container inventory_management_system_image
```
If you want to pass values for the environment variables then instead run:
```
docker run -p 80:80 --name inventory_management_system_container --env API_URL=/api inventory_management_system_image
```

## Testing instructions
- [ ] Review code
- [ ] Check Actions build
- [ ] GitHub Actions should build and push an image to Harbor

## Agile board tracking
closes #45

### CODE REVIEW

**Comment by VKTB** [comment]
The optimised production build now failing with a different error - https://github.com/ral-facilities/inventory-management-system/issues/140.

**Comment by joshuadkitenge** [comment]
> The optimised production build now failing with a different error - #140.

https://github.com/ral-facilities/operationsgateway/blob/develop/craco.config.js


```
      webpackConfig.output.clean = {
          keep(asset) {
            // exclude mockServiceWorker.js from build
            return !asset.includes('mockServiceWorker.js');
          },
        };
```

Adding this may solve the problem

**Comment by VKTB** [comment]
> > The optimised production build now failing with a different error - #140.
> 
> https://github.com/ral-facilities/operationsgateway/blob/develop/craco.config.js
> 
> ```
>       webpackConfig.output.clean = {
>           keep(asset) {
>             // exclude mockServiceWorker.js from build
>             return !asset.includes('mockServiceWorker.js');
>           },
>         };
> ```
> 
> Adding this may solve the problem

Just a note to say that we tried this and it did not work.

**Comment by joelvdavies** [comment]
I

> > > The optimised production build now failing with a different error - #140.
> > 
> > 
> > https://github.com/ral-facilities/operationsgateway/blob/develop/craco.config.js
> > ```
> >       webpackConfig.output.clean = {
> >           keep(asset) {
> >             // exclude mockServiceWorker.js from build
> >             return !asset.includes('mockServiceWorker.js');
> >           },
> >         };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Adding this may solve the problem
> 
> Just a note to say that we tried this and it did not work.

I had a brief look at it - I think that code is necessary to prevent the `mockServiceWorker.js` in the public folder from ending up in the build so we do need it anyway, but the issue appears to be the import itself here https://github.com/ral-facilities/inventory-management-system/blob/05614225d928bc3b01342d79c68981d79db3c883/src/index.tsx#L171 & all of the mock files in `src/mocks` that imports it in someway. It appears the build still looks it up regardless of whether the `prepare` method is called or not (I deleted everything below it but left the function uncalled - which worked when I had also added `rm -rf src/mocks` to the docker RUN command to remove the mock files to it). Perhaps there is some build option that can ignore things like this like the tests when doing a production build?

**Comment by louise-davies** [comment]
@joelvdavies I did raise to Josh that I also encountered issues with the existing `prepare` function in OG as well and showed it to him, he probably didn't get a chance to test it though. The improved code I did is here: https://github.com/ral-facilities/operationsgateway/blob/cfe38273cb929ae9df9a6a00969c7c06aaa5f0b4/src/index.tsx#L179

Feel free to try this out and see if it fixes things. The webpack config just slightly optimises the build and makes it smaller.

**Comment by VKTB** [comment]
@louise-davies I remember Joshua trying this but it was still failing with the same error.

**Comment by joelvdavies** [comment]
> I don't think the `REACT_APP_INVENTORY_MANAGEMENT_SYSTEM_BUILD_DIRECTORY` variable is properly being set following the recent commits, I am investigating.

You're right, I didn't realise sed didn't save the file 🤦‍♂️

File: .dockerignore
Comment (VKTB) [inline .dockerignore:4]:
```suggestion
**/build
**/coverage
**/cypress
**/lib
**/node_modules
```

**Comment by VKTB** [comment]
@joelvdavies The changes you made look good to me. It can be merged in after you approve it.

**Review by joelvdavies — APPROVED** [review]
LGTM

**Pull request merged**
